### ISSUE TITLE
[iOS] Wrong routing in the new flutter version (3.13.1)

### ISSUE THREAD

**Issue by Primo85** [flutter/flutter#135418]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Open test app // https://github.com/Primo85/testFlutterApp
2. Navigate to home(or history) screen
3. Open other app // https://github.com/Primo85/flutterCaller
4. Click "go to ourApp://some.domain.com"

### Expected results

In the version of Flutter we are currently using(3.10.1), our test application is opened in the location where we left it(home or history screen) and with the state we left it in.

### Actual results

In the new version of Flutter(3.13.1), our test application is opened in the mainScreen as if we had used 'ourApp://some.domain.com/'.

### Code sample

Our test application in which we reproduced the problem: 
https://github.com/Primo85/testFlutterApp

The iOS helper app that we use to open our test app:
https://github.com/Primo85/flutterCaller

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[!] Flutter (Channel stable, 3.13.1, on macOS 13.1 22C65 darwin-x64, locale en-PL)
    ! Warning: `dart` on your path resolves to /usr/local/Cellar/dart/2.18.7/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/pbiskup/fvm/versions/3.13.1. Consider adding /Users/pbiskup/fvm/versions/3.13.1/bin to the front of your path.
[!] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.
[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.1)
[✓] VS Code (version 1.81.1)
[✓] Connected device (3 available)
[✓] Network resources

! Doctor found issues in 2 categories.
```

</details>

**Comment by Primo85** [comment]
Maybe my test notes will be helpful:
<img width="1096" alt="Screenshot 2023-09-25 at 14 36 07" src="https://github.com/flutter/flutter/assets/23534114/fbe04c90-e521-46f7-9859-f5d13b9192e8">

**Comment by darshankawar** [comment]
Thanks for the report @Primo85 
I see below changes made in 3.13 version that you can check and see if it relates to the behavior you are seeing:

https://github.com/flutter/flutter/pull/126453
https://github.com/flutter/flutter/pull/119968

**Comment by Primo85** [comment]
I'm currently investigating this matter and I already have a place in the code where the problem occurs. I'm still thinking about a solution...

**Comment by Primo85** [comment]
Hi @darshankawar
In version 3.13.1, there is new code on line 65 of the lib/src/widgets/router.dart file:
```dart
String get location {
    if (_location != null) {
      return _location!;
    }
    return Uri.decodeComponent(
      Uri(
        path: uri.path.isEmpty ? '/' : uri.path,
        queryParameters: uri.queryParametersAll.isEmpty ? null : uri.queryParametersAll,
        fragment: uri.fragment.isEmpty ? null : uri.fragment,
      ).toString(),
    );
  }
```
I replaced line 71 with:
```dart
path: uri.path.isEmpty ? '' : uri.path,
```
... and the code works as we expected.
In case ourApp://some.domain.com our test application is opened in the location where we left it. 
In case ourApp://some.domain.com/ our test application is opened in the mainScreen.

**Comment by darshankawar** [comment]
Thanks for the analysis @Primo85 
Can you take a look at [this](https://github.com/flutter/flutter/issues/123985#issuecomment-1694652493) and see if it helps to answer / resolve your issue ?

**Comment by Primo85** [comment]
Hi @darshankawar.
This is not the case. We don't use apple-app-site-association, we open our app from another app.

**Comment by darshankawar** [comment]
Thanks for the update. Using the code sample provided in [original description](https://github.com/flutter/flutter/issues/135418#issue-1911392985), and also the details mentioned [here](https://github.com/flutter/flutter/issues/135418#issuecomment-1765913283), I am able to replicate the reported behavior.

This seems more like a navigator related issue than go_router package one, so labeling accordingly.

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.13.6, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.13.6 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ead455963c (5 days ago), 2023-09-26 18:28:17 -0700
    • Engine revision a794cf2681
    • Dart version 3.1.3
    • DevTools version 2.25.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.16.0-14.0.pre.23, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.16.0-14.0.pre.23 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision eefafdacdb (53 minutes ago), 2023-10-20 00:06:34 -0400
    • Engine revision 9dd59f7fcf
    • Dart version 3.3.0 (build 3.3.0-45.0.dev)
    • DevTools version 2.28.1
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/dhs/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 65.1.4
    • Dart plugin version 213.7228

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 15.3.1 19D52
    • macOS (desktop)           • macos                                    •
      darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 109.0.5414.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by goderbauer** [comment]
cc @chunhtai

**Comment by chunhtai** [comment]
The previous behavior seems like a bug to me. When you open an app with deep link, it will pass the route to the app. In this case '' is translate into '/' which I think is semantically correct?

**Comment by Primo85** [comment]
Thank you @chunhtai 
What if we want to keep the previous behavior in the application? That means opening it in exactly the same place where we left it.

**Comment by chunhtai** [comment]
The flutter app needs to ignore the deeplink request. I look at the code there isn't a way to handle deeplink differently in gorouter yet. One thing you can try is to register a WidgetsBindingObserver before the app is run

```
void main() {
  runApp(
    IgnoreDeeplink(
      child: MaterialApp.router(routerConfig: _router)
    )
  );
}

IgnoreDeeplink extends StatefulWidget {...}

IgnoreDeeplinkState extends State<IgnoreDeeplink> with WidgetsBindingObserver {
....
  @override
  Future<bool> didPushRouteInformation(RouteInformation routeInformation) async => true

}
```

**Comment by chunhtai** [comment]
If you have more question please use the discord channel hackers-routing

**Comment by Primo85** [comment]
Hi @chunhtai,
I tried using didPushRouteInformation as you suggested, but it didn't work. This method was not called in fact at all.
After looking at FlutterAppDelegate.mm I decided to work around the problem in method `application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool` and not call super.method for a specific case.
What is your suggestion to be able to open an iOS app in a previous view with a previous state and still be able to support deeplinks?
